python_modules/dagster/dagster/_core/definitions/assets.py

```diff
@@ -1286,27 +1290,32 @@ def update_replace_dict_and_conflicts(
 
     def map_asset_specs(self, fn: Callable[[AssetSpec], AssetSpec]) -> "AssetsDefinition":
         mapped_specs = []
+        additional_deps_per_key: Dict[AssetKey, Set[AssetDep]] = {}
         for spec in self.specs:
             mapped_spec = fn(spec)
             if mapped_spec.key != spec.key:
                 raise DagsterInvalidDefinitionError(
                     f"Asset key {spec.key.to_user_string()} was changed to "
                     f"{mapped_spec.key.to_user_string()}. Mapping function must not change keys."
                 )
-            if (
-                # check reference equality first for performance
-                mapped_spec.deps is not spec.deps and mapped_spec.deps != spec.deps
-            ):
-                raise DagsterInvalidDefinitionError(
-                    f"Asset deps {spec.deps} were changed to {mapped_spec.deps}. Mapping function "
-                    "must not change deps."
-                )
+
+            deps_per_key = defaultdict(list)
+            for dep in mapped_spec.deps:
+                deps_per_key[dep.asset_key].append(dep)
+            for dep in spec.deps:
+                if dep.asset_key not in deps_per_key:
+                    raise DagsterInvalidDefinitionError(
+                        f"After mapping, asset {dep.asset_key.to_user_string()} was not found in the mapped dependencies. You should not remove dependencies from assets, since that can destroy the internal structure of the asset."
```

Comment:
hmmmm why not let the user do this? I'm not really following why this should always be illegal

Reply:
Hmmm that does not match up to the description in the error message. If this is for  arguemnt-based deps that should be explained in the errror message. However this should work for vanilla deps.

Reply:
I'll have to dig into whether there's a way to disambiguate argument based deps from non-argument based deps.

Reply:
I was able to find a way to disambiguate, so I think we're good to support this 👍